From 3277fda98135c2c9e83f592e84a1202251ad457a Mon Sep 17 00:00:00 2001
From: Jared Duke <jdduke@google.com>
Date: Fri, 12 Aug 2022 14:38:06 -0700
Subject: [PATCH 01/30] Explicitly madvise boot image oat files

This restores a subset of previous behavior for madvising all oat
files, specifically for boot image oat files used for execution.
This can greatly improve ZygoteInit performance for class loading.

Bug: 239467545
Test: m + inspect madvise logs
Change-Id: I31deb6b0e57f4e3a5497eda944031280759da999
---
 runtime/gc/space/image_space.cc | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/runtime/gc/space/image_space.cc b/runtime/gc/space/image_space.cc
index 6afd63e4a5..f29e6bbfa8 100644
--- a/runtime/gc/space/image_space.cc
+++ b/runtime/gc/space/image_space.cc
@@ -3111,8 +3111,24 @@ class ImageSpace::BootImageLoader {
         return false;
       }
     }
+
+    // As an optimization, madvise the oat file into memory if it's being used
+    // for execution with an active runtime. This can significantly improve
+    // ZygoteInit class preload performance.
+    if (executable_) {
+      Runtime* runtime = Runtime::Current();
+      if (runtime != nullptr) {
+        Runtime::MadviseFileForRange(runtime->GetMadviseWillNeedSizeOdex(),
+                                     oat_file->Size(),
+                                     oat_file->Begin(),
+                                     oat_file->End(),
+                                     oat_file->GetLocation());
+      }
+    }
+
     space->oat_file_ = std::move(oat_file);
     space->oat_file_non_owned_ = space->oat_file_.get();
+
     return true;
   }
 
-- 
2.42.0

